nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -877,6 +878,120 @@ impl<'a> Planner<'a> {
+                )
+                .all(|zone| inventory_zones.contains_key(&zone.id))
+            {
+                info!(
+                    self.log, "zones not yet up-to-date";
+                    "sled_id" => %sled_id,
+                );
+                return Ok(());
+            }
+        }
+
+        // Choose among the out-of-date zones one with the fewest dependencies.
+        let mut out_of_date_zones = sleds
+            .into_iter()
+            .flat_map(|sled_id| {
+                self.blueprint
+                    .current_sled_zones(
+                        sled_id,
+                        BlueprintZoneDisposition::is_in_service,
+                    )
+                    .filter_map(|zone| {
+                        (zone.image_source
+                            != self
+                                .blueprint
+                                .zone_image_source(zone.zone_type.kind()))
+                        .then(|| (sled_id, zone.clone()))
+                    })
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<(SledUuid, BlueprintZoneConfig)>>();
+
+        sort_zones_by_deps(&mut out_of_date_zones);
+        if let Some((sled_id, zone)) = out_of_date_zones.first() {
+            return self.update_or_expunge_zone(*sled_id, zone);
+        }
+
+        info!(self.log, "all zones up-to-date");
+        Ok(())
+    }
+
+    /// Update a zone to use a new image source, either in-place or by
+    /// expunging it and letting it be replaced in a future iteration.
+    fn update_or_expunge_zone(
+        &mut self,
+        sled_id: SledUuid,
+        zone: &BlueprintZoneConfig,
+    ) -> Result<(), Error> {
+        let zone_kind = zone.zone_type.kind();
+        let image_source = self.blueprint.zone_image_source(zone_kind);
+        if zone.image_source == image_source {
+            // This should only happen in the event of a planning error above.
+            warn!(
+                self.log, "zone is already up-to-date";
+                "sled_id" => %sled_id,
+                "zone_id" => %zone.id,
+                "kind" => ?zone.zone_type.kind(),
+                "image_source" => %image_source,
+            );
+        } else {
+            match zone_kind {
+                ZoneKind::Crucible
+                | ZoneKind::CockroachDb
+                | ZoneKind::Clickhouse => {
+                    info!(
+                        self.log, "updating zone image source in-place";
+                        "sled_id" => %sled_id,
+                        "zone_id" => %zone.id,
+                        "kind" => ?zone.zone_type.kind(),
+                        "image_source" => %image_source,
+                    );
+                    self.blueprint.sled_set_zone_source(
+                        sled_id,
+                        zone.id,
+                        image_source,
+                    )?;
+                }
+                _ => {
```

Comment:
I think we should spell out all the `ZoneKind`s, so if additional variants are added in the future we are forced to make a decision here.

Reply:
Good idea, also done in 050933ef7.